src/diagnostics.rs

```diff
@@ -57,35 +57,58 @@ pub fn publish_diagnostics(params: Params, ctx: &mut Context) {
             )
         })
         .join(" ");
-    let mut lines_with_errors = HashSet::new();
-    let diagnostic_ranges = diagnostics
+
+    // Assemble a list of diagnostics by line number
+    let mut diagnostic_ranges = HashMap::new();
+    for x in diagnostics {
+        let line_number = x.range.end.line;
+
+        let face = match x.severity {
+            Some(DiagnosticSeverity::Error) => "InlayDiagnosticError",
+            _ => "InlayDiagnosticWarning",
+        };
+
+        let line_diagnostics = diagnostic_ranges
+            .entry(line_number)
+            .or_insert(LineDiagnostics {
+                range_end: x.range.end,
+                symbols: String::new(),
+                text: x
+                    .message
+                    .split('\n')
+                    .next()
+                    .unwrap_or_default()
+                    .replace("|", "\\|"),
+                text_face: face.to_string(),
+            });
+
+        line_diagnostics
+            .symbols
+            .push_str(&format!("{{{}}}%opt[lsp_inlay_diagnostic_sign]", face))
+    }
+
+    // Assemble ranges based on the lines
+    let diagnostic_ranges = diagnostic_ranges
         .iter()
-        .map(|x| {
-            let face = match x.severity {
-                Some(DiagnosticSeverity::Error) => "InlayDiagnosticError",
-                _ => "InlayDiagnosticWarning",
-            };
-            // Pretend the language server sent us the diagnostic past the end of line
-            let line = x.range.end.line;
-            let line_text = get_line(line as usize, &document.text);
-            let mut pos =
-                lsp_position_to_kakoune(&x.range.end, &document.text, ctx.offset_encoding);
+        .map(|(line_number, line_diagnostics)| {
+            let line_text = get_line(*line_number as usize, &document.text);
+            let mut pos = lsp_position_to_kakoune(
+                &line_diagnostics.range_end,
+                &document.text,
+                ctx.offset_encoding,
+            );
             pos.column = line_text.len_bytes() as u32;
-            // separate all but the first diagnostic on the same line
-            let sep = if lines_with_errors.insert(line) {
-                ""
-            } else {
-                ", "
-            };
-            editor_quote(&format!(
-                "{}+0|{{{}}}{{\\}}{} {}",
+
+            format!(
+                "\"{}+0|%opt[lsp_inlay_diagnostic_gap]{} {{{}}}{{\\}}{}\"",
```

Comment:
We'll get an error if `%opt{lsp_inlay_diagnostic_gap}` contains a `|`.
I guess we don't care but maybe there's an easy fix.

Reply:
I added a note specifying that `|` isn't allowed.